dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/CommittingPartitionTracker.java

```diff
@@ -0,0 +1,304 @@
+import java.util.concurrent.TimeUnit;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.KafkaAdapterFactory;
+import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.OffsetAndMetadata;
+import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.TopicPartition;
+
+/**
+ * Tracks the assignment of a partition to this consumer and commits message offsets to that partition
+ * <p>
+ * In addition to the function of {@link PartitionTracker}, this class additionally commits message offsets back to the kafka broker in response to a call to
+ * {@link #recordDone(long, Optional)}.
+ * <p>
+ * For performance, each completed record isn't committed immediately. Instead, this class tries to batch up completed records and commit them together. There are several
+ * parameters that control how often the message offset is committed:
+ * <ul>
+ * <li>{@code maxCommitBatchSize}: sets the maximum number of records to wait for before committing the offset</li>
+ * <li>{@code maxCommitBatchInterval}: sets the maximum time to wait after a {@link #recordDone(long, Optional)} is called before committing the offset</li>
+ * </ul>
+ */
+public class CommittingPartitionTracker extends PartitionTracker {
+
+    private static final TraceComponent tc = Tr.register(CommittingPartitionTracker.class);
+
+    private final KafkaAdapterFactory factory;
+    private final ScheduledExecutorService executor;
+    private final KafkaInput<?, ?> kafkaInput;
+    private final int maxCommitBatchSize;
+    private final Duration maxCommitBatchInterval;
+
+    /**
+     * Set of CompletedWork which has either not been committed or has been committed but the commit has not completed yet.
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private final SortedSet<CompletedWork> completedWork = new TreeSet<>();
+
+    /**
+     * The count of CompletedWork for which a commit has not been started
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private int outstandingUncommittedWork = 0;
+
+    /**
+     * The last offset for which a commit was started
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private long committedOffset;
+
+    /**
+     * The currently scheduled task which will attempt to commit completed work or {@code null} if no task has been scheduled.
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private Future<?> pendingCommitTask = null;
+
+    /**
+     *
+     * @param topicPartition the partition to track
+     * @param factory the KafkaAdaptorFactory
+     * @param kafkaInput the KafkaInput
+     * @param initialCommittedOffset the position of the reader when the partition was assigned
+     * @param executor a ScheduledExecutorService
+     * @param maxCommitBatchSize the maximum number of records to wait for before committing the offset
```

Comment:
Does a value less than or equal to zero actually mean unlimited batch size?

Reply:
Yes, I'll add doc for zero being a special value for both `maxCommitBatchSize` and `maxCommitBatchInterval`.

Reply:
zero or less